torch.quantization conversion utilities, observers for eager mode quantization

jerryzh168 opened the pull request:
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#22010 torch.quantization conversion utilities, observers for eager mode quantization**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D15554183/)

torch.quantization module with observers and conversion routines

Differential Revision: [D15554183](https://our.internmc.facebook.com/intern/diff/D15554183/)

jerryzh168 committed:
V10: add quantized Tensor comparison in common_utils.py

Differential Revision: D15375695
Differential Version: 85144759

jerryzh168 committed:
V12: Initial commit

Differential Revision: D15483071
Differential Version: 85154440

jerryzh168 committed:
V11: comment

Differential Revision: D15375695
Differential Version: 85155178

jerryzh168 committed:
V12: added test

Differential Revision: D15554224
Differential Version: 85162142

jerryzh168 committed:
V14: Merge with parent diff changes

Differential Revision: D15483071
Differential Version: 85187565

jerryzh168 committed:
V15: (no description)

Differential Revision: D15483071
Differential Version: 85213062

jerryzh168 committed:
V13: Merge with parent diff changes

Differential Revision: D15554224
Differential Version: 85213957

jerryzh168 committed:
V12: update docs

Differential Revision: D15375695
Differential Version: 85219380

jerryzh168 committed:
V16: docs

Differential Revision: D15483071
Differential Version: 85220005

jerryzh168 committed:
V14: add future import

Differential Revision: D15554224
Differential Version: 85217753

jerryzh168 committed:
V15: docs

Differential Revision: D15554224
Differential Version: 85223788

jerryzh168 committed:
V23: Initial commit

Differential Revision: D15554183
Differential Version: 85267282

jerryzh168 committed:
V24: (no description)

Differential Revision: D15554183
Differential Version: 85267665

jerryzh168 committed:
V17: (no description)

Differential Revision: D15483071
Differential Version: 85322492

jerryzh168 committed:
V18: Merge with parent diff changes

Differential Revision: D15554224
Differential Version: 85325488

jerryzh168 committed:
V19: (no description)

Differential Revision: D15554224
Differential Version: 85330125

jerryzh168 committed:
V18: (no description)

Differential Revision: D15483071
Differential Version: 85331722

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,240 @@
+            # Obtain quantized weights and biases
+            observerConfig = mod.qConfig['weight']['config']
+            weight_observer = mod.qConfig['weight']['observer'](observerConfig)
+            quantized_weight, quantized_bias = quantizeWeightAndBias(
```
why this logic can't sit inside nnq.Linear constructor? it'd be much more sensible. Is there a case where we want to create a raw nnq.Linear? I doubt so. Then you don't need all this set_state business

raghuramank100 replied:
Yes, we may want to create a raw nnq.Linear. This is important as someone may want to quantize modules their own way (not using this framework) and still be able to access the quantized kernels. This way of splitting the functionality allows for that.

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,240 @@
+            }
+            new_mod.set_state(state)
+        if type(mod) == torch.nn.Identity:
+            new_mod = nnq.Quantize(mod.qparams[0].item(), mod.qparams[1].item(), torch.quint8)
```
can we standardize interface for nnq module inputs? For example - always just accept an original module. Then the constructor can poke into qparams, original weights, etc. And you don't need a fancy swapModule function - it can just take the dictionary of module remappings and operate generically

raghuramank100 replied:
We had originally started with that, but moved to this approach. The reasoning is that one can create a quantized module from scratch and come up with any scheme for creating quantized weights and calculating scale/zero-points and still be able to run it in pytorch.  Having the conversion coupled with the quantized module forces us to mandate one way of doing the quantization, which is too restrictive

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,47 @@
+import torch
+
+class Observer(nn.Module):
+    def __init__(self, config):
```
comments and specify/verify type of config

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,47 @@
+
+        return qparams
+
+class WeightObserver(Observer):
```
why do we even need weight observer? is it for QAT? because for PTQ you do know the weight values and can just access them in the constructor of nnq module

raghuramank100 replied:
Strictly, we do not. This was done to keep the API for weights and activations consistent, instead of introducing something different. For weights the observer is stateless and gets applied at init. It is also consistent with how quantization is done, we reduce the weights into a histogram (bin them) and then compute scale and zero-point from that.

dzhulgakov reviewed (commented):
some comments (I feel I already provided some of these on Raghu's prototype but I'm not sure where it was)

dzhulgakov commented on test/test_quantization.py:
```diff
@@ -0,0 +1,207 @@
+        qconfig = {}
+        qconfig_activation = tq.QConfig(torch.qint8, 'per_tensor_symmetric')
+        qconfig_weight = tq.QConfig(torch.qint8, 'per_tensor_symmetric')
+        qconfig['weight'] = {'observer': tq.WeightObserver,
```
`observer_factory(torch.qint8, 'per_tensor_symmetric') -> ObserverClass`

```
ObserverInstance.dtype()
ObserverInstance.qscheme()
```

dzhulgakov commented on test/test_quantization.py:
```diff
@@ -0,0 +1,207 @@
+        qconfig_weight = tq.QConfig(torch.qint8, 'per_tensor_symmetric')
+        qconfig['weight'] = {'observer': tq.WeightObserver,
+                             'config': qconfig_weight}
+        qconfig['activation'] = {'observer': tq.Observer,
```
maybe qconfig to be namedtuple?

jerryzh168 committed:
V14: Merge with parent diff changes

Differential Revision: D15375695
Differential Version: 85439410

jerryzh168 committed:
V19: Merge with parent diff changes

Differential Revision: D15483071
Differential Version: 85439430

jerryzh168 committed:
V25: Merge with parent diff changes

Differential Revision: D15554183
Differential Version: 85439445

jerryzh168 committed:
V20: Merge with parent diff changes

Differential Revision: D15554224
Differential Version: 85439441

jerryzh168 committed:
V26: (no description)

Differential Revision: D15554183
Differential Version: 85439488

dzhulgakov commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -159,3 +159,14 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
                                       missing_keys, unexpected_keys, error_msgs)
         return
+
+    def set_state(self, state):
```
seems to be not used now (since you put things in from_float)

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+    for module in self.children():
+        module.forward(output)
+
+def addObserver(myMod, qConfigDict, qConfigParent=None, prefix=''):
```
what is our python naming scheme? I was under impression that it's names_with_underscores (please correct me if I'm wrong)

jerryzh168 replied:
Yeah I think we should use that, will change.

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+        prefix: corresponding prefix of the current module, used as key in
+                qConfigDict
+    """
+    myMod.qConfig = None
```
I'd check that module already doesn't have this attribute

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+from .observer import *
+
+def _forward_hook(self, input, output):
+    for module in self.children():
```
wait, why is this the forward hook? why it can't just explicitly call .observer ?

jerryzh168 replied:
this is for modifying the behavior of forward function, isn't that needed?

dzhulgakov replied:
meaning why are we iterating over .children()? can't you just do `self.observer(output)`?

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+        myMod.register_forward_hook(_forward_hook)
+
+    for name, child in myMod.named_children():
+        if name is not 'observer':
```
you add observer only if there are no children, so you can remove it and just put the whole loop into for

also `is not 'observer'` doesn't do what you think it does: https://stackoverflow.com/questions/2988017/string-comparison-in-python-is-vs

jerryzh168 replied:
I think the reason is adding observer changed 'myMod.named_children()' it will go into infinite loop without this check..

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+    # Insert observers only for leaf nodes
+    if myMod.qConfig is not None and len(myMod._modules) == 0:
+        myMod.add_module('observer', myMod.qConfig.activation())
+        myMod.register_forward_hook(_forward_hook)
```
do you remove this hook later when quant/dequant are inserted?

jerryzh168 replied:
yeah we should..

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+    Returns:
+        new_fn: a modified function which calls quant -> base_bn -> dequant
+    """
+    def new_fn(x):
```
did you try the approach we discussed with creating a derived class?

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+    if name in qConfigDict:
+        return name
+    elif name:
+        parent = '.'.join(name.split('.')[:-1])
```
`name.rsplit('.',1)[0]` is more idiomatic

dzhulgakov replied:
comment still here

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+def getModule(modInstance, name):
+    splits = name.split('.')
+    mod = modInstance._modules[splits[0]]
+    for i in range(1, len(splits)):
```
`for split in splits[1:]:`

and there's no reason to separate out the first iteration of the loop

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+            mod_key_list.append(dict_key)
+
+    for name in set(mod_key_list):
+        if name is not '':
```
just make `getModule(modInstance, '')` work

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+
+def addQuantDeQuant(modInstance, qConfigDict):
+    r"""Add AbstractQuant and AbstractDeQuant module and modify forward function
+    to call quant dequant accordign to qConfigDict
```
you need to give longer explanation here. In particular, it seems that you insert quant/dequant only at modules specified in the config (and not their children), is that right? my original though was that you'd quantize everything within that subtree and 'a.b.c' naming mechanism is just to pass the same setting to the entire module subtree.

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+    """
+    def __init__(self, qconfig):
+        super(AbstractQuant, self).__init__()
+        self.add_module('observer', qconfig.activation())
```
why do you have observer here if you already have an observer on parent module. Something doesn't add up here

jerryzh168 replied:
removed.

jerryzh168 replied:
actually we need this for input activations I think.

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+    if hasattr(MyModule, 'observer'):
+        if isinstance(MyModule.observer, Observer):
+            # Simple symmetric quantization over entire observed range
+            MyModule.qparams = MyModule.observer.calculate_qparams()
```
why do we need to assign it on module level and create more pollution? why calculate_qparams can't be called during the conversion process?

jerryzh168 replied:
yeah I think we can move these to the place when we need them.

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+    step and observers will record the stats of the tensors, after that we will
+    call `convert` which will convert the module to a quantized module.
+    """
+    prepare(module, qConfigDict)
```
I wonder whether `quantize()` overall is even needed. It might be fine to just ask user to explicitly call prepare/convert. I'd rename `convert` into something different then

jerryzh168 replied:
don't we need to have eval_fn as well?
I think it's just a wrapper over these three functions listed below and it gets us a one-line API for quantizing a model.

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+}
+
+ABSTRACT_MODULE_MAPPING = {
+    AbstractQuant: nnq.Quantize,
```
QuantStub or QuantSpot or something like that might be a better name

jerryzh168 replied:
changed to Stub

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+        new_mod = ABSTRACT_MODULE_MAPPING[type(mod)].from_float(mod)
+
+    # keep the modification to forward
+    new_mod.forward = mod.forward
```
wait, why do we need to do that? quant / dequant should be already be gone by now, no?

if you want to preserve this logic - it'd be much more meaningful to put calls to quant/dequant explicitly in the parent module's forward(). Relying on `forward` overrides is hacky in itself

jerryzh168 replied:
removed.

dzhulgakov commented on torch/quantization/QConfig.py:
```diff
@@ -0,0 +1,10 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+from collections import namedtuple
+
+class QOptions(object):
```
I'd drop this method all together. AFAIU, it's only used to pass arguments to constructor of observer. You can just allow to pass arbitrary arguments there (including dtype/qscheme directly).

jerryzh168 replied:
OK, yeah I feel this is not really used as well.

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,62 @@
+        return x
+
+def observer(observer_cls, qoptions):
+    def get_instance():
```
you can just use `functools.partial`

dzhulgakov commented on test/test_quantization.py:
```diff
@@ -0,0 +1,189 @@
+        myModel = Model1()
+        qConfigDict = {
+            # Top level qconfig
+            '': QConfig(weight=observer(WeightObserver, QOptions(torch.qint8, torch.per_tensor_affine)),
```
I think what makes sense to expose is `default_weight_observer(torch.qint8, torch.per_tensor_affine)`. And I'd make per_tensor.affine a default option

dzhulgakov reviewed (changes requested):
getting there, but there's still a lot to improve in terms of clarity of code

jerryzh168 commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,221 @@
+        mod: module to be changed
+        qConfig: quantization configuration
+
+    """
```
return WrappedModule(mod)

class WrappedModule(Module):
          def __init__(self, mod):
                 quant = AbstractQuant(qconfig)
                 dequant = AbstractDeQuant()

          def forward(self, X):
                 self.quant(X)
                 self.mod.forward(X)
                 self.dequant(X)

jerryzh168 replied:
I think the disadvantage of this is it changes the module hierarchy, not sure whether this will be a problem.

jerryzh168 committed:
V27: (no description)

Differential Revision: D15554183
Differential Version: 85527513

jerryzh168 committed:
V28: (no description)

Differential Revision: D15554183
Differential Version: 85527974

jerryzh168 committed:
V29: (no description)

Differential Revision: D15554183
Differential Version: 85563533

dzhulgakov commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -36,7 +36,8 @@ def forward(self, X):
     @staticmethod
     def from_float(mod):
-        return Quantize(mod.qparams[0].item(), mod.qparams[1].item(), torch.quint8)
+        qparams = mod.observer.calculate_qparams()
```
better error messaging? if `.observer` is not present - produce an error that the module hasn't been instrumented

dzhulgakov commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -36,7 +36,8 @@ def forward(self, X):
     def from_float(mod):
-        return Quantize(mod.qparams[0].item(), mod.qparams[1].item(), torch.quint8)
+        qparams = mod.observer.calculate_qparams()
+        return Quantize(qparams[0].item(), qparams[1].item(), torch.quint8)
```
shouldn't it be something like Observer.dtype? don't we want to make dtype configurable?

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,61 @@
+    arguments after that. and it should provide a calculate_qparam function.
+    TODO: Maybe add an abstract Observer class?
+    """
+    def __init__(self, **kwargs):
```
why do you have this weird kwargs business? regular arguments with defgault values? (and docs please!)

jerryzh168 replied:
reason for kwargs is this is a base class and other types of observers might have some extra arguments. but I think we should have some default arguments like dtype and qscheme as well.

dzhulgakov replied:
wait, how base class is related here. If one of the descendants overrides this constructor, it'd pass the subset of argument down. There's no need for kwargs

jerryzh168 replied:
OK

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,61 @@
+        self.qscheme = kwargs.get('qscheme', torch.per_tensor_affine)
+        assert self.qscheme == torch.per_tensor_affine, \
+            'Default Observer only works for per_tensor_affine quantization scheme'
+        self.stats = torch.tensor([-1, 1], dtype=torch.float)
```
is starting from -1/1 desired?

jerryzh168 replied:
maybe too small, [-6, 6]?

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,61 @@
+
+    def calculate_qparams(self):
+        if self.dtype == torch.qint8:
+            qparams = torch.zeros(2).float()
```
nit: can't you pull out everything except for `qparams[1]` out of the if?

jerryzh168 replied:
yes..

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,61 @@
+    Other types of Observers should follow the same API, the first argument of
+    the constructor should be qoptions, and it can take arbitrary number of
+    arguments after that. and it should provide a calculate_qparam function.
+    TODO: Maybe add an abstract Observer class?
```
TODO shouldn't be in the docstring but in the comment (we don't want to pollute docs)

dzhulgakov replied:
And having it as a abstract class is indeed a good idea

raghuramank100 replied:
Do we need an abstract class or a default class that everyone needs to inherit from (like nn.Module). Default observer would have a reference get stats/calcQparams, users can inherit and update.

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,61 @@
+        self.stats = torch.tensor([torch.min(x), torch.max(x)], dtype=torch.float)
+        return x
+
+def observer(observer_cls, kwargs):
```
I'd duplicate the args here to have good documentation (and you can also move a check on qscheme being only one to here)

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+    for module in self.children():
+        module.forward(output)
+
+def add_observer(module, qconfig_dict, qconfig_parent=None, prefix=''):
```
it's probably useful to expose this method as part of public API too. If so - keep the add_observer with just two args and have a recursive function to be a helper function (so that we don't expose prefix and qconfig_parent as part of docs)

dzhulgakov replied:
still not exposed as part of public api? let's do that

raghuramank100 replied:
I agree, so this method just adds an observer to a single module, correct?

raghuramank100 replied:
I agree, so this method just adds an observer to a single module, correct?

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+
+    # Insert observers only for leaf nodes, note that this observer is for
+    # the output of the module, for input QuantStub will observe them
+    if module.qconfig is not None and len(module._modules) == 0:
```
no need to access `_modules`, do `len(module.children())`

dzhulgakov replied:
not addressed

jerryzh168 replied:
I checked this one, `module.children()` is an iterator, we'll have to do `len(list(module.children())` should we do that instead of accessing `_modules`?

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+    if module.qconfig is not None and len(module._modules) == 0:
+        # observer will be when we swap the module
+        module.add_module('observer', module.qconfig.activation())
+        # when do we remove hooks?
```
you probably want a pair function that does removing observers

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+    for module in self.children():
+        module.forward(output)
+
+def add_observer(module, qconfig_dict, qconfig_parent=None, prefix=''):
```
the great debate - add_observer or add_observers? :)

jerryzh168 replied:
I'll use `add_observer`, we'll have `add_quant_dequant` as well

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+            module_prefix = prefix + '.' + name if prefix else name
+            add_observer(child, qconfig_dict, module.qconfig, module_prefix)
+
+def add_quant_dequant_call(base_fn):
```
you can delete it now?

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+        return x
+    return new_fn
+
+class QuantWrapper(nn.Module):
```
docstring

also, does it produce a reasonable repr(m)?

jerryzh168 replied:
we will have extra hierarchy in the leaf node.

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+        X = self.dequant(X)
+        return X
+
+def get_config_key(name, qconfig_dict):
```
nit, can you rewrite it as a while loop? might be much more readable

jerryzh168 replied:
I feel recursion is always more readable :) but sure

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+        return mod
+    splits = name.split('.')
+    for split in splits:
+        mod = mod._modules[split]
```
_modules is internal, don't use internal fields. Either getattr or named_children(). (getattr is faster probably)

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+    assert name != ''
+    qconfig = qconfig_dict[name]
+    splits = name.split('.')
+    print('splits:', splits)
```
delete?

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+        if dict_key is not None:
+            mod_key_set.add(dict_key)
+
+    # We need to update the modules from inner most
```
I feel it'd be more readable as a recursive function that just modifies the module and return either it or the new wrapper. Then you don't need `get_config_key` and can just traverse the dictionary at the same time (kind of like you do in add_observer)

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+        parent_module = module
+        module = module._modules[split]
+    assert parent_module is not module
+    parent_module._modules[splits[-1]] = QuantWrapper(module, qconfig)
```
again, just do setattr

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+    print('splits:', splits)
+    for split in splits:
+        parent_module = module
+        module = module._modules[split]
```
getattr

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+            add_quant_dequant_module(module, qconfig_dict, name)
+    return module
+
+def prepare(module, qconfig_dict):
```
should it be called `observe_and_quantize`?

raghuramank100 replied:
It is not quantizing here, only preparing for quantization. Also, if we want the same names for QAT, better to leave it at prepare: In the QAT case, prepare would insert fake-quants.

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+    return add_quant_dequant(module, qconfig_dict)
+
+class QuantStub(nn.Module):
+    r"""Quantize stub module which will be replace to actual Quantize module
```
expand to what's the purpose of it is? E.g. "represents future quantization point in the model. QuantStub would collect input statistic when running the float model and later can be swapped with quantized.Quantize module"

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+    """
+    def __init__(self, qconfig):
+        super(QuantStub, self).__init__()
+        self.add_module('observer', qconfig.activation())
```
to continue this discussion - what happens if I manually insert QuantStub in my model? The add_observer would fail since it already has an observer. We probably want to handle QuantStub separately there or we can reverse the order of add_quant_dequant and adding observers

raghuramank100 replied:
Or we could have the add_observer function check if a module already has an observer and do nothing in that case

dzhulgakov replied:
I think it you put quant_dequant addition step (either automatic or manual) before add_observers, then you don't need to have special handling of observer here - it'd be just magically handled by injection code

raghuramank100 replied:
Yes, that is possible: Right now the quantstub has an observer within it, we will need to remove that for it to work.

jerryzh168 replied:
quant_dequant changes the hierarchy so we can't swap the order

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+    step and observers will record the stats of the tensors, after that we will
+    call `convert` which will convert the module to a quantized module.
+    """
+    module = prepare(module, qconfig_dict)
```
it's a bit confusing that we both return the modified module tree and mutate the original one. Not sure how to fix it consistently (we could do deepcopy, but I'm not sure it's worth it)

jerryzh168 replied:
the step that modifies it is `add_quant_dequant`, all other functions are modifying module inplace. I think we cannot avoid returning new module if we use `QuantWrapper` because the top level module can be replaced with `QuantWrapper(module)`, however, if we rule out this possibility, i.e. constraint user to only provide a non-terminal module as input, then we don't need to return.

dzhulgakov replied:
I'd add it to docstring that module would be modified in place but the new wrapper can be returned

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,222 @@
+    convert_to_quantized(module)
+    return module
+
+def convert_to_quantized(module):
```
mapping should be the argument here as well (as that's the top api)

dzhulgakov replied:
not addressed

raghuramank100 replied:
By mapping you mean the dictionary of module -> quant modules ? I feel that should not be specified by the user: As we add support for more nn.quantized modules, shouldnt it get automatically updated?  User can anyway control which module gets quantized by qconfigdict.

dzhulgakov replied:
What if I define my own module though? We can provide the default map (which should be a default value), but it should be overridable imho

jerryzh168 replied:
@raghuramank100 should be fine, user can override default by DEFAULT + {...}

dzhulgakov commented on test/test_quantization.py:
```diff
@@ -0,0 +1,276 @@
+def eval_fn(model, data):
+    model(data)
+
+default_options = {
```
btw, should it be just globally default options for the Observer call?

jerryzh168 replied:
yeah, added to observer.py

dzhulgakov commented on test/test_quantization.py:
```diff
@@ -0,0 +1,276 @@
+        model = SingleLinearModel()
+        qConfigDict = {
+            # Top level qconfig
+            '': QConfig(weight=observer(WeightObserver, default_options),
```
should there be a default QConfig thingy that just returns a pair weight/activations with default arguments? (and maybe allow to override dtype / qscheme for them). It's a bit verbose to spell it out like this

dzhulgakov commented on test/test_quantization.py:
```diff
@@ -0,0 +1,276 @@
+                           activation=observer(Observer, default_options))
+        }
+        eval_args = [calib_data]
+        myModel = tq.quantize(myModel, qConfigDict, eval_fn, *eval_args)
```
why do you need `eval_args` like this?

jerryzh168 replied:
don't know.. I'll change to list.

dzhulgakov commented on test/test_quantization.py:
```diff
@@ -0,0 +1,276 @@
+        }
+        eval_args = [calib_data]
+        myModel = tq.prepare(myModel, qConfigDict)
+        # self.assertEqual(hasattr(myModel, 'quant'), False)
```
why all of these tests commented out?

jerryzh168 replied:
because I added QuantWrapper in this diff.

dzhulgakov commented on torch/quantization/__init__.py:
```diff
@@ -0,0 +1,10 @@
+from .observer import *  # noqa: F401
+from .QConfig import *  # noqa: F401
+
+_all__ = [
```
these names need updating

dzhulgakov reviewed (changes requested):
Apart from top comments, there are a lot of minor nits that would be easy to spot if you just read through entire code at once (I recommend to do that)

jerryzh168 committed:
V1: Initial commit

Differential Revision: D16022491
Differential Version: 85687669

jerryzh168 committed:
V2: Merge with parent diff changes

Differential Revision: D16022491
Differential Version: 85728558

jerryzh168 committed:
V3: Merge with parent diff changes

Differential Revision: D16022491
Differential Version: 85811636

jerryzh168 committed:
V31: test cleanup is wip, but convert_modules.py is mostly ready for review

Differential Revision: D15554183
Differential Version: 85770232

jerryzh168 committed:
V32: (no description)

Differential Revision: D15554183
Differential Version: 85843886

dzhulgakov commented on test/test_quantization.py:
```diff
@@ -0,0 +1,254 @@
+        myModel = NestedModel()
+        qConfigDict = {
+            'fc3': default_qconfig,
+            'sub2': default_qconfig
```
I'm still a bit uneasy about using these settings to control non-inserting quant/dequant. There are a few problems with it:

- It's non-obvious that `forward` of the module needs to be runnable with quantized values. It's very easy to miss some functional call there and have wtf moment
- the property that particular module can be wrapped as quant/dequant is a property of module **type** and not module **instance**. So it'd be more natural to have a separate configuration that lists which models support "entire module" quantization or just enable it for people who want to do it manually
- I'd interpret qConfig applied to a parent module (e.g. "sub2") to be just observer settings to be applied to all modules in the subgraph that are supported and **skip** module types that are unknown. Otherwise we're signing ourselves up for spelling every individual sub-module explicitly if I just want to get behavior similar to torch/jit/quantized.py which is swapping out all modules that support quantization.

Here's a different proposal - we have two types of settings:
- qconfig that describes observer settings only
- list of module types that can be quantized whole-sale. By default they are the keys from DEFAULT_MODULE_MAPPING list but I one can add new settings.

Then .prepare would receive both of those setting and apply quant/dequant insertion only for modules of supported types who are in the parts of the tree present in qconfig.

The graph mode api would just receive qconfig directly as the compiler knows best which modules are "supported" and where to insert quantization blocks (though if we're thinking of customization, we might pass some per-module-type settings too).

Thoughts?

raghuramank100 replied:
Good points,  I agree with some of them:
1. Great point about forward. We have no way of knowing if a module's forward consists of quantizable operations, short of doing tracing. So the default module swapping behavior should replace a module with q->quant-module->dq or fq->quant-module->fq in the QAT case. If the user is sure that a given module is fully quantizable, then we could have a 'strict=True' option that would not insert quant dequants. This will not be the default.
2. I think having quantization controllable at the level of a module instance is important. Think of resnet-101, in this case we have 101 layers that are identical, (nn.conv2d), but we do want to be able to control which layers we want to quantize. It is also important for the automatic precision selection work as we can figure out which parts of a network we want to quantize.
3. Agreed with (3). My understanding of (3) is that if I have a module for which I have a qconfig, I would attempt to quantize every sub-module within it. If I find a submodule that is not supported (say a softmax), that would not be modified. Note that using strict=True would dump an error message in this case.
4. We are planning on a quantization debugger (proposal is a quip doc) that allows for whitelist/blacklist) of ops etc as a bootcamp/HAM task.
For graph mode, the API would remain the same, the user would get the benefits of fusion (better insertion of quant/dequant), no extra quant/dequants (as we know the entire execution graph).

dzhulgakov replied:
2. I'm fully agreeing that we want to have control of quantization on module instance level. I was talking only about inserting quant/dequant. So in ResNet101 case I can still specify qconfig on pieces of modules. But if I want to say that module MyAwesomeModule has a good "forward" such that quantizing everything inside it is sufficient to just put quant/dequant around it - we should specify it by passing 'MyAwesomeModule''s type, not instance as it's safe to apply to all instances that are covered in qconfig

3. Agree with your reasoning. But as outlined above I'd decouple qconfig for quant/dequant insertion a bit

More examples of what exactly I mean. Let's say we have a hierarchy:
```
m : MyModel
  sub1 : MyModule
    fc : FC
    conv: Conv
  sub2 : MyModule
    fc : FC
    conv: Conv
  sub3 : MyOtherModule
    fc1 : FC
    fc2 : FC
```

Let's say MyModule has good forward and MyOtherModule has a "bad" one.

Then `qconfig={"": ...}` would quantize all individual FC/CONV layers and insert quant/dequant around each of them.

If I pass `qconfig={"": ...}, whole_modules={MyModule}`, then quant/dequant would be inserted around `sub1, sub2, sub3.fc1, sub3.fc2`

If I pass `qconfig={"sub2": ...}, whole_modules={MyModule}`, then quant/dequant would be inserted around `sub2` only (sub1 is not in qconfig).

If I pass `qconfig={"sub2": default_qconfig, "sub2.fc": None}, whole_modules={MyModule}`, then quant/dequant would be inserted around `sub2.conv` only - because part of MyModule is not quantized according to qconfig.

Does it make sense?

Also for more complex quant/dequant insertions in eager mode - we should really just ask people to manually insert QuantStub modules into them.

raghuramank100 replied:
I agree, this makes more sense than the 'strict=True' that I was proposing.  I was only defining default behaviors in the previous comments: 
- Default would correspond to whole_modules={} and we would insert q-dq per module that we can convert. 
- If whole_module is specified, we would error out if MyModule had something within that is not quantizable at preparation time itself, If there is an operation in forward that is not quantizable (say a sqrt for example), the error would be seen only at run-time as the assumption of the whole_module being quantizable is violated.

jerryzh168 replied:
makes sense, I'll add module setting.

jerryzh168 replied:
maybe a different PR, this is out for too long.

dzhulgakov commented on torch/quantization/__init__.py:
```diff
@@ -0,0 +1,16 @@
+    for data in calib_data:
+        model(data)
+
+_all__ = [
```
where did all the names go? :)

jerryzh168 replied:
will add them after API is finalized.

dzhulgakov reviewed (commented):
There are many earlier comments that are still not addressed

raghuramank100 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -37,7 +37,9 @@ def forward(self, X):
 
     @staticmethod
     def from_float(mod):
```
Should we have this API instead be:
from_float(mod, weight_qparams, activation_qparams)
mod is a nn module: Without observers, qconfig
Benefits:  Allows for users to still use from_float and specify directly the qparams without using our mechanism of observers etc

jerryzh168 replied:
OK

raghuramank100 commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,183 @@
+
+    return module_swapped
+
+DEFAULT_MODULE_MAPPING = {
```
Related to my previous comment: These need not be arguments to the quantize function, can instead be dictionaries like here.

jerryzh168 committed:
V34: use QuantWrapper, need cleanup

Differential Revision: D15554183
Differential Version: 85937790

jerryzh168 committed:
V35: (no description)

Differential Revision: D15554183
Differential Version: 85937931

dzhulgakov commented on test/test_quantization.py:
```diff
@@ -0,0 +1,273 @@
+import torch.quantization as tq
+from torch.quantization import *
+from common_utils import TestCase, run_tests
+
```
please also add a test that explicitly inserts QuantStub/DequantStub and doesn't call add_quant_dequant. This is to make sure that we don't force everyone into that API

jerryzh168 replied:
added a test

dzhulgakov commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -37,7 +37,9 @@ def forward(self, X):
     @staticmethod
     def from_float(mod):
-        return Quantize(mod.qparams[0].item(), mod.qparams[1].item(), torch.quint8)
+        assert hasattr(mod, 'observer'), 'Input float module must have observer attached'
```
shouldn't we just enforce the type to QuantStub?

dzhulgakov commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -140,3 +142,39 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+        if qparams_dict:
+            assert('activation' in qparams_dict)
+            assert('weight' in qparams_dict)
+            act_qparams = qparams_dict['activation']
```
why not just pass activation_qparams / wt_qparams directly as args?

also, how will this branch work for non-linear quantization? you just pass qparams, not the quantizer. Should it receive Quantizer in the future? (if quantizer is not ready - I'd rather hide this API, it's always possible to add it later)

jerryzh168 replied:
sure we can remove this branch for now

dzhulgakov commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -140,3 +142,39 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
+            weight_observer = mod.qconfig.weight()
+            weight_observer(mod.weight)
+            wt_qparams = weight_observer.calculate_qparams()
+        bias_qparams = torch.zeros(2)
```
not used?

dzhulgakov commented on torch/quantization/QConfig.py:
```diff
@@ -0,0 +1,19 @@
+from collections import namedtuple
+from .observer import *
+
+default_weight_qoptions = {
```
this and the next one not used?

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,149 @@
+        # observer and hook will be gone after we swap the module
+        module.add_module('observer', module.qconfig.activation())
+        module.register_forward_hook(_observer_forward_hook)
+        return QuantWrapper(module, module.qconfig)
```
wait, didn't we agree to separate observer insertion and quant_dequant? what if I just want to insert QuantStub on my own? One way to structure add_quant_dequant API is to operate only on modules supported in REMAPPING and whose names are in qconfig subtree

jerryzh168 replied:
Oh right, I'll split

raghuramank100 replied:
So, is the plan to have two functions:
 - Add observer that only inserts observers
 - Insert Q DQ that would insert Q and DQ stubs before/after each module within a specified sub-module in QconfigDict.
If whole module is true, then  inserting Q/DQ should happen only at the beginning and end of each sub-module.

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,149 @@
+    r"""Converts the float module with qparams to a quantized module.
+    """
+    module_swapped = swap_module(module)
+    if len(list(module.named_children())) == 0:
```
no need in this return - you can just run code below, it'd do nothing

jerryzh168 committed:
V36: please take a look at code, docs and tests are not ready yet

Differential Revision: D15554183
Differential Version: 85997154

raghuramank100 commented on torch/quantization/convert_modules.py:
```diff
@@ -4,23 +4,45 @@
-        self.dequant = DeQuantStub()
-        self.module = module
+def propogate_qconfig_helper(module, qconfig_dict, qconfig_parent=None, prefix=''):
+    r"""This is a helper function for `propogate_qconfig`
```
NIT: propagate_qconfig

raghuramank100 commented on torch/quantization/convert_modules.py:
```diff
@@ -36,26 +58,48 @@ def add_observer(module, qconfig_dict, qconfig_parent=None, prefix=''):
+        return QuantWrapper(module, module.qconfig)
+
+    for name, child in module.named_children():
+        module._modules[name] = add_quant_dequant_helper(child)
```
This function is missing?

jerryzh168 replied:
oh, will rename, we don't need the helper function

raghuramank100 commented on torch/quantization/convert_modules.py:
```diff
@@ -69,7 +113,9 @@ def prepare(module, qconfig_dict):
-    return add_observer(module, qconfig_dict)
+    propogate_qconfig(module, qconfig_dict)
+    add_observer(module)
+    return add_quant_dequant(module)
```
We need to add two additional options in qconfig_dict per module:  quantize_input, dequantize_output. 
add_quant_dequant would look at the qconfig of each module and decide whether to insert quant/dequant.
This complicates the qconfigDict specification: For example, if I want to quantize the whole model I would need to say:

qConfigDict={' ': { weight: weight_qconfig, activation:activation_qconfig, quantize_input:True, dequantize_output: True}}
For each of the child modules, assuming all of them are quantized, I would want to specify quantize_input=False and dequantize_output = False. 

One way to solve this issue is to have default for quantize_input = False and dequantize_output = False. If user wants to override it, then qConfigDict needs to provided explicitly.

jerryzh168 replied:
I don't feel we need to do that, user can modify their model directly if they want to have the whole_module functionality. they can skip the first step of inserting quant_dequant and start with add_observers

jerryzh168 replied:
adding this to config is complicates things a lot and does buy much convenience

raghuramank100 replied:
We will have a hierarchy mismatch in this case:
For example, lets have:

`class TwoLayerLinearModel(torch.nn.Module):
    def __init__(self):
        super(TwoLayerLinearModel, self).__init__()
        self.fc1 = torch.nn.Linear(5, 8).to(dtype=torch.float)
        self.fc2 = torch.nn.Linear(8, 5).to(dtype=torch.float)

    def forward(self, x):
        x = self.fc1(x)
        x = self.fc2(x)
        return x


myModel = TwoLayerLinearModel()
tq.prepare(myModel)
print(myModel)
# Would print:
fc1
 Quant
 module
 DeQuant
fc2
 Quant
 module
 DeQuant

#If the user does this manually, by inserting quant/dequant: we would have:
Quant
fc1
fc2
DeQuant
'


self.fc1.module, otherwise, it will show up as: self.fc1.

jerryzh168 replied:
we can ask user to use QuantWrapper to insert quant dequant as well?

jerryzh168 replied:
QuantWrapper might not cover all the cases, in case of more complicated topologies I think it is OK to have different hierarchy since our tool will not work in that case which means there is no corresponding quantized module using our tool.

dzhulgakov replied:
I concur with what Jerry said - we shouldn't overcomplicate this tooling.

As for hierarchy mismatch - **it's ok, you're modifying your model**. Note that state_dict won't cause problems as noone would be loading checkpoints from float version directly.

As for qconfig - since it's recursive it'd work just fine - `fc1` would apply also for `fc1.module`

raghuramank100 replied:
I agree, had a chat with Jerry to discuss this further. Lets keep it simple.  insert quant dequant will only insert quant/dequant for every module, no need to have other functionality.

raghuramank100 replied:
> I concur with what Jerry said - we shouldn't overcomplicate this tooling.
> 
> As for hierarchy mismatch - **it's ok, you're modifying your model**. Note that state_dict won't cause problems as noone would be loading checkpoints from float version directly.
> 
> As for qconfig - since it's recursive it'd work just fine - `fc1` would apply also for `fc1.module`

For QAT, we would be loading state_dict as we train and save checkpoints, but still changing the hierarchy is ok, so I agree we should keep the API simple

raghuramank100 commented on torch/quantization/convert_modules.py:
```diff
@@ -69,7 +113,9 @@ def prepare(module, qconfig_dict):
                       configuration
     """
-    return add_observer(module, qconfig_dict)
+    propogate_qconfig(module, qconfig_dict)
```
I think we should flip the order here:
Zero: Propagate QconfigInfo
First call add_quant_dequant: This inserts quant and dequant stubs - no observers anywhere. QuantStub in this case is a pure placeholder does not contain observers.
Second: Insert observers: This would insert observers for both QuantStubs and for the module itself.
User is free to skip the add_quant_dequant step and do it manually. Again, QuantStub does not contain observers.

jerryzh168 replied:
sounds good

dzhulgakov replied:
yes, we should

raghuramank100 commented on torch/quantization/convert_modules.py:
```diff
@@ -125,17 +174,12 @@ def convert_to_quantized(module):
 }
 
-def swap_module(mod, mapping=DEFAULT_SWAP_MODULE_MAPPING):
+def swap_module(mod, mapping):
```
Can we combine STUB_MODULE_MAPPING and DEFAULT_MODULE_MAPPING into one?

jerryzh168 replied:
after changing the order I think we can do that, since all modules will have observer

raghuramank100 reviewed (changes requested):
Please see comments, thanks

dzhulgakov commented on test/test_quantization.py:
```diff
@@ -0,0 +1,296 @@
+class ManualQuantModel(torch.nn.Module):
+    def __init__(self):
+        super(ManualQuantModel, self).__init__()
+        self.quant = QuantStub(default_qconfig)
```
if we treat QuantStub same as other modules - should we not receive qconfig here? (or make it optional)

jerryzh168 replied:
in our tool when we insert QuantStub the qconfig propagation is finished, so we need to provide a qconfig. I can make it optional, when qconfig is not provided, QuantStub will not have a `qconfig` field so that propagate_qconfig can go through.

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,193 @@
+        propogate_qconfig_helper(child, qconfig_dict, module.qconfig, module_prefix)
+
+def propogate_qconfig(module, qconfig_dict):
+    r"""Propogate qconfig through the module hierarchy given a qconfig dict
```
it's not a very good description, how about:

```
Recursively propagate qconfig through the module hierarchy and assign `qconfig` attribute on each leaf module

...

 qconfig_dict: dictionary that maps from name of submodule to quantization
                     configuration. Qconfig applies to all submodules of a given module unless lower-level entries are specified

```

and propAgate :)

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,193 @@
+    self.observer(output)
+
+def add_observer(module):
+    r"""Transforms a module to a quantized module according to qconfig_dict.
```
outdated description?

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,193 @@
+        if name != 'observer':
+            add_observer(child)
+
+def get_config_key(name, qconfig_dict):
```
not used?

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,193 @@
+    r"""Add QuantStub and DeQuantStub module and Wrap the module in QuantWrapper
+    according to qconfig_dict
+    """
+    is_leaf = len([name for (name, _) in module.named_children() if name != 'observer']) == 0
```
you can use `any` instead of `len` - it'd stop after first match

jerryzh168 replied:
since order is changed we don't need this anymore, but thanks for the tip

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,193 @@
+def prepare(module, qconfig_dict):
+    r"""Prepares the module for calibration according to qconfig_dict.
+
+    The function adds observer and quant dequant stub modules to the module.
```
to every leaf module that has qconfig specified in the dictionary for itself or for one of its parents

jerryzh168 replied:
I'll remove this part since these are documented in functions called by this one.

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,193 @@
+    r"""Prepares the module for calibration according to qconfig_dict.
+
+    The function adds observer and quant dequant stub modules to the module.
+    It also adds the forward hooks to call quant and dequant methods.
```
this is not true anymore?

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,70 @@
+
+    def __init__(self, dtype=torch.qint8, qscheme=torch.per_tensor_affine):
+        super(WeightObserver, self).__init__(dtype, qscheme)
+        self.stats = torch.tensor([-6, 6], dtype=torch.float)
```
is -6 some magic number? :) just set it to None so that something fails unless forward() is called?

jerryzh168 replied:
OK

dzhulgakov commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,70 @@
+        assert self.qscheme == torch.per_tensor_affine, \
+            'Default Observer only works for per_tensor_affine quantization scheme'
+        # Symmetric range for initialization
+        self.stats = torch.tensor([-6, 6], dtype=torch.float)
```
btw, why are we doing exponential smoothing? I'd imagine explicitly calculating some percentile of distribution would be much more meaningful default. The problem with exponential smoothing is that it biases towards last batches and also -6/6 doesn't make sense.

@raghuramank100 ?

raghuramank100 replied:
Agreed, lets go with min/max , just like we do for weights.

dzhulgakov reviewed (changes requested):
I think structure-wise we're very close. Please do another pass for docstrings and small details

jerryzh168 committed:
V37: (no description)

Differential Revision: D15554183
Differential Version: 86033432

jerryzh168 committed:
V38: (no description)

Differential Revision: D15554183
Differential Version: 86036343

jerryzh168 committed:
V39: more docs and tests

Differential Revision: D15554183
Differential Version: 86040531

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,244 @@
+    """
+    self.observer(output)
+
+def add_observer(module):
```
I wonder whether we should have also `remove_observer`. The thing is that right now you might attach observers to some modules that you don't swap out later. So might make sense to clean up. (for removing hooks - you can store hook inside observer class and then do .remove() on it)

jerryzh168 replied:
OK, I'll add a TODO for this

dzhulgakov commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,244 @@
+    """
+    def __init__(self, module):
+        super(QuantWrapper, self).__init__()
+        assert hasattr(module, 'qconfig'), 'Please add qconfig to module before \
```
Wait, why is that? you can just make QuantWrapper insertion **before** propagate_qconfig and everything should work as long as qwrappers are on the leaf nodes.

Probably best is to just check whether qconfig is present on module and then copy it, otherwise keep it empty

dzhulgakov reviewed (commented):
Yay! It looks good to me. I'll let @raghuramank100 do a pass and answer question about exponential smoothing. But I think it's good to go

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,70 @@
+        self.stats = torch.tensor([torch.min(x), torch.max(x)], dtype=torch.float)
+        return x
+
+def observer(observer_cls, **kwargs):
```
Do we need these three functions? Why not directly use the Observer class?

jerryzh168 replied:
we need these because we need to create a new instance of observer every time. so qconfig should contain something that can create a observer instance.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,70 @@
+
+        return qparams
+
+class WeightObserver(Observer):
```
We could delete WeightObserver as it is now identical to Observer

jerryzh168 replied:
weight is just one batch of data, but activation is multiple batches of data how do we do that using just min/max?

raghuramank100 reviewed (changes requested):
Almost there, a few more comments.

jerryzh168 committed:
V40: observers

Differential Revision: D15554183
Differential Version: 86050037

raghuramank100 commented on torch/quantization/convert_modules.py:
```diff
@@ -51,6 +51,7 @@ def _observer_forward_hook(self, input, output):
     """
     self.observer(output)
 
+# TODO(jerryzh): remove_observer?
```
Not sure if we really need it. We can instead show in the tutorial how to do a deepcopy, so that the original float module is still available.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,60 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
```
Lets go with observers that support symmetric quant for weights and asymmetric quant for activations. Also, lets go with just taking the min and max: clipping the weights can cause significant errors.  This is the default, we will add more observers in subsequent PRs.  A baseline observer can be the following:    
`def __init__(self, dtype=torch.quint8, qscheme=torch.per_tensor_symmetric):
        super(Observer, self).__init__()
        self.dtype = dtype
        self.qscheme = qscheme
        assert (self.qscheme in (torch.per_tensor_affine,torch.per_tensor_symmetric)  \
            'Default Observer only works for per_tensor quantization schemes'
    

    def forward(self, x):
        self.min_val = torch.min(x).float()
        self.max_val = torch.max(x).float()

    def calculate_qparams(self):
        if self.dtype == torch.qint8:
            qmin, qmax = -128, 127
        else:
            qmin, qmax = 0, 255
      if self.q_scheme == torch.per_tensor_symmetric:
                max_val = torch.max(-self.min_val, self.max_val)
                min_val = -max_val*(128.0/127.0)
        n_levels = 255.0
        scale = (max_val - min_val) / n_levels
        zero_point = qmin - round(min_val / scale)
        zero_point = max(qmin, zero_point)
        zero_point = min(qmax, zero_point)
    
    return torch.tensor([scale, zero_point])

jerryzh168 replied:
symmetric quant is not supported in quantized kernels now?

raghuramank100 replied:
It is, symmetric quant is the default, asymmetric quant support is being added.

jerryzh168 committed:
V41: Merge with parent diff changes

Differential Revision: D15554183
Differential Version: 86079795

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,60 @@
+def observer(observer_cls, **kwargs):
+    return partial(observer_cls, **kwargs)
+
+def default_observer(**kwargs):
```
For activations, we need to override the defaults:
```
kwargs.setdefault('dtype', torch.quint8)
kwargs.setdefault('qscheme', torch.per_tensor_affine)
```

raghuramank100 reviewed (changes requested):
Please see comments, thanks

jerryzh168 committed:
V43: observers

Differential Revision: D15554183
Differential Version: 86093005

jerryzh168 committed:
V44: (no description)

Differential Revision: D15554183
Differential Version: 86094049

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -0,0 +1,68 @@
+        n_levels = 255.0
+        max_val, min_val = self.max_val.item(), self.min_val.item()
+        if self.qscheme == torch.per_tensor_symmetric:
+            max_val = max(-min_val, max_val)
```
NIT: Lets simplify to:
```
  scale = max_val/127.0
 ```
remove calculation of min_val

raghuramank100 commented on torch/quantization/convert_modules.py:
```diff
@@ -0,0 +1,244 @@
+    module = add_quant_dequant(module)
+    add_observer(module)
+    return module
+
```
Make add_quant_dequant optional: This allows for quantize (the top level API) to work for cases where the user manually inserts quantstubs or calls quantwrapper.
```
def prepare(module, qconfig_dict, enable_add_quant_dequant=True):
    propagate_qconfig(module, qconfig_dict)
   if enable_add_quant_dequant:
        module = add_quant_dequant(module)
    add_observer(module)
```
    return module

jerryzh168 replied:
moving add_quant_dequant as discussed.

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -0,0 +1,352 @@
+    def test_quant_wrapper(self):
+        model = WrappedModel()
+
+        tq.propagate_qconfig(model)
```
Can replace this test and other tests with call to tq.quantize(), with the suggested change below

raghuramank100 reviewed (changes requested):
Add an option to prepare: i.e enable/disable add_quant_dequant. This allows for tq.quantize to work for all the test cases:   including quantwrapper and manual quantstub insertion.

jerryzh168 committed:
V45: (no description)

Differential Revision: D15554183
Differential Version: 86106388

raghuramank100 reviewed (changes requested):
Just one more test requested. Thanks!

jerryzh168 commented:
@raghuramank100 what test? I have removed add_quant_dequant from default and changed the previous test to call everything explicitly. The quantize api are used in the last few tests(manual and quantwrapper test cases)

jerryzh168 committed:
V46: (no description)

Differential Revision: D15554183
Differential Version: 86168758

raghuramank100 reviewed (approved):
Looks great!

ezyang commented:
This broke link https://travis-ci.org/pytorch/pytorch/jobs/556423310
